File: src/memory/knowledge/chunking.ts
import { extname } from 'node:path';

import {
  chunkContentHash,
  generateChunkId,
  type Chunk,
  type ChunkOptions,
} from './types.js';

const DEFAULT_TARGET_SIZE = 1600;
const DEFAULT_OVERLAP_SIZE = 320;

/** Check if content looks binary (contains null bytes). */
function isBinary(content: string): boolean {
  return content.includes('\0');
}

/**
 * Split content into logical sections based on file type.
 * Returns arrays of line groups, where each group is an array of
 * { lineNumber (1-indexed), text } objects.
 */
function splitIntoSections(
  fileLines: string[],
  ext: string,
): { lineNumber: number; text: string }[][] {
  if (ext === '.md' || ext === '.rst') {
    return splitMarkdown(fileLines);
  }
  if (['.ts', '.tsx', '.js', '.jsx', '.py'].includes(ext)) {
    return splitCode(fileLines);
  }
  // Plain text: split on double newlines (paragraph boundaries)
  return splitParagraphs(fileLines);
}

/**
 * Split markdown into sections at H2+ headers and paragraph boundaries.
 * Keeps fenced code blocks intact.
 */
function splitMarkdown(
  fileLines: string[],
): { lineNumber: number; text: string }[][] {
  const sections: { lineNumber: number; text: string }[][] = [];
  let current: { lineNumber: number; text: string }[] = [];
  let inCodeBlock = false;

  for (let i = 0; i < fileLines.length; i++) {
    const line = fileLines[i];
    const lineObj = { lineNumber: i + 1, text: line };

    // Track fenced code blocks
    if (line.trimStart().startsWith('```')) {
      inCodeBlock = !inCodeBlock;
      current.push(lineObj);
      continue;
    }

    // Split on H2+ headers when not inside code block
    if (!inCodeBlock && /^#{2,}\s/.test(line) && current.length > 0) {
      sections.push(current);
      current = [lineObj];
      continue;
    }

    // Split on blank lines (paragraph boundary) when not in code block
    // A blank line after non-blank content marks a paragraph break
    if (
      !inCodeBlock &&
      line.trim() === '' &&
      current.length > 0 &&
      current.some((l) => l.text.trim() !== '')
    ) {
      // Include this blank line in current section, then start fresh
      current.push(lineObj);
      sections.push(current);
      current = [];
      continue;
    }

    current.push(lineObj);
  }

  if (current.length > 0) {
    sections.push(current);
  }

  return sections;
}

/** Split code at blank lines between top-level definitions. */
function splitCode(
  fileLines: string[],
): { lineNumber: number; text: string }[][] {
  const sections: { lineNumber: number; text: string }[][] = [];
  let current: { lineNumber: number; text: string }[] = [];

  for (let i = 0; i < fileLines.length; i++) {
    const line = fileLines[i];
    const lineObj = { lineNumber: i + 1, text: line };

    if (line.trim() === '' && current.length > 0) {
      // Check if the previous non-blank line ended a block
      // and the next non-blank line starts a new one
      const nextNonBlank = fileLines.slice(i + 1).find((l) => l.trim() !== '');
      if (nextNonBlank !== undefined) {
        sections.push(current);
        current = [lineObj];
        continue;
      }
    }

    current.push(lineObj);
  }

  if (current.length > 0) {
    sections.push(current);
  }

  return sections;
}

/** Split plain text on paragraph boundaries (blank lines). */
function splitParagraphs(
  fileLines: string[],
): { lineNumber: number; text: string }[][] {
  const sections: { lineNumber: number; text: string }[][] = [];
  let current: { lineNumber: number; text: string }[] = [];

  for (let i = 0; i < fileLines.length; i++) {
    const line = fileLines[i];
    const lineObj = { lineNumber: i + 1, text: line };

    if (line.trim() === '' && current.length > 0) {
      sections.push(current);
      current = [lineObj];
      continue;
    }

    current.push(lineObj);
  }

  if (current.length > 0) {
    sections.push(current);
  }

  return sections;
}

/** Get text from a section (array of line objects). */
function sectionText(section: { text: string }[]): string {
  return section.map((l) => l.text).join('\n');
}

/**
 * Chunk a file into semantic pieces with overlap.
 *
 * @param filePath - Relative path from repo root
 * @param content - File text content
 * @param options - Chunking options (targetSize, overlapSize)
 * @returns Array of Chunk objects
 */
export function chunkFile(
  filePath: string,
  content: string,
  options?: ChunkOptions,
): Chunk[] {
  // Empty or whitespace-only
  if (content.trim() === '') return [];

  // Binary detection
  if (isBinary(content)) return [];

  const targetSize = options?.targetSize ?? DEFAULT_TARGET_SIZE;
  const overlapSize = options?.overlapSize ?? DEFAULT_OVERLAP_SIZE;

  const fileLines = content.split('\n');
  const ext = extname(filePath).toLowerCase();

  const sections = splitIntoSections(fileLines, ext);

  // Merge small sections until reaching targetSize, then emit a chunk
  const chunks: Chunk[] = [];
  let accumulated: { lineNumber: number; text: string }[] = [];
  let accumulatedLength = 0;

  function emitChunk(
    lines: { lineNumber: number; text: string }[],
    overlapLines: { lineNumber: number; text: string }[],
  ): { lineNumber: number; text: string }[] {
    if (lines.length === 0) return [];

    const allLines = [...overlapLines, ...lines];
    const text = allLines.map((l) => l.text).join('\n');
    const startLine = allLines[0].lineNumber;
    const endLine = allLines[allLines.length - 1].lineNumber;

    chunks.push({
      id: generateChunkId(filePath, startLine, endLine),
      filePath,
      startLine,
      endLine,
      text,
      contentHash: chunkContentHash(text),
    });

    // Compute overlap: take lines from the end of `lines` that fit overlapSize
    if (overlapSize <= 0) return [];
    const overlapResult: { lineNumber: number; text: string }[] = [];
    let overlapLen = 0;
    for (let i = lines.length - 1; i >= 0; i--) {
      const lineLen = lines[i].text.length + 1; // +1 for newline
      if (overlapLen + lineLen > overlapSize && overlapResult.length > 0) break;
      overlapResult.unshift(lines[i]);
      overlapLen += lineLen;
    }
    return overlapResult;
  }

  let overlapLines: { lineNumber: number; text: string }[] = [];

  for (const section of sections) {
    const sectionLen = sectionText(section).length;

    // If adding this section exceeds targetSize and we have accumulated content, emit
    if (accumulatedLength > 0 && accumulatedLength + sectionLen > targetSize) {
      overlapLines = emitChunk(accumulated, overlapLines);
      accumulated = [];
      accumulatedLength = 0;
    }

    accumulated.push(...section);
    accumulatedLength += sectionLen;

    // If this single section alone exceeds targetSize, emit it immediately
    if (accumulatedLength > targetSize) {
      overlapLines = emitChunk(accumulated, overlapLines);
      accumulated = [];
      accumulatedLength = 0;
    }
  }

  // Emit remaining accumulated content
  if (accumulated.length > 0) {
    emitChunk(accumulated, overlapLines);
  }

  return chunks;
}

File: go/internal/knowledge/chunking.go
// Package knowledge provides document chunking, indexing, embedding, and search.
package knowledge

import (
	"crypto/sha256"
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
)

const (
	DefaultTargetSize  = 1600
	DefaultOverlapSize = 320
)

// SupportedExtensions lists file extensions that can be chunked.
var SupportedExtensions = map[string]bool{
	".md": true, ".txt": true, ".rst": true,
	".ts": true, ".tsx": true, ".js": true, ".jsx": true, ".py": true,
}

// codeExtensions are extensions that use code-style section splitting.
var codeExtensions = map[string]bool{
	".ts": true, ".tsx": true, ".js": true, ".jsx": true, ".py": true,
}

// Chunk represents a piece of a documentation file.
type Chunk struct {
	ID          string
	FilePath    string
	StartLine   int
	EndLine     int
	Text        string
	ContentHash string
}

// ChunkOptions controls chunking behavior.
type ChunkOptions struct {
	TargetSize  int
	OverlapSize int
}

// GenerateChunkID creates a deterministic ID from file path and line range.
func GenerateChunkID(filePath string, startLine, endLine int) string {
	h := sha256.Sum256([]byte(fmt.Sprintf("%s:%d:%d", filePath, startLine, endLine)))
	return fmt.Sprintf("%x", h[:8])
}

// ChunkContentHash computes SHA-256 of text content.
func ChunkContentHash(text string) string {
	h := sha256.Sum256([]byte(text))
	return fmt.Sprintf("%x", h)
}

// lineObj tracks a line's number and text.
type lineObj struct {
	lineNumber int
	text       string
}

var headerRegex = regexp.MustCompile(`^#{2,}\s`)

// ChunkFile splits a file into semantic chunks with overlap.
func ChunkFile(filePath, content string, opts *ChunkOptions) []Chunk {
	if strings.TrimSpace(content) == "" {
		return nil
	}
	if strings.Contains(content, "\x00") {
		return nil
	}

	targetSize := DefaultTargetSize
	overlapSize := DefaultOverlapSize
	if opts != nil {
		if opts.TargetSize > 0 {
			targetSize = opts.TargetSize
		}
		if opts.OverlapSize > 0 {
			overlapSize = opts.OverlapSize
		}
	}

	fileLines := strings.Split(content, "\n")
	ext := strings.ToLower(filepath.Ext(filePath))

	sections := splitIntoSections(fileLines, ext)

	var chunks []Chunk
	var accumulated []lineObj
	accumulatedLength := 0
	var overlapLines []lineObj

	for _, section := range sections {
		sectionLen := sectionTextLen(section)

		// Emit if accumulated + section exceeds target and we have content
		if accumulatedLength > 0 && accumulatedLength+sectionLen > targetSize {
			overlapLines = emitChunk(filePath, accumulated, overlapLines, overlapSize, &chunks)
			accumulated = nil
			accumulatedLength = 0
		}

		accumulated = append(accumulated, section...)
		accumulatedLength += sectionLen

		// Emit if single section exceeds target
		if accumulatedLength > targetSize {
			overlapLines = emitChunk(filePath, accumulated, overlapLines, overlapSize, &chunks)
			accumulated = nil
			accumulatedLength = 0
		}
	}

	// Emit remaining
	if len(accumulated) > 0 {
		emitChunk(filePath, accumulated, overlapLines, overlapSize, &chunks)
	}

	return chunks
}

func emitChunk(filePath string, lines, overlapLines []lineObj, overlapSize int, chunks *[]Chunk) []lineObj {
	if len(lines) == 0 {
		return nil
	}

	allLines := append(overlapLines, lines...)
	text := joinLines(allLines)
	startLine := allLines[0].lineNumber
	endLine := allLines[len(allLines)-1].lineNumber

	*chunks = append(*chunks, Chunk{
		ID:          GenerateChunkID(filePath, startLine, endLine),
		FilePath:    filePath,
		StartLine:   startLine,
		EndLine:     endLine,
		Text:        text,
		ContentHash: ChunkContentHash(text),
	})

	// Compute overlap from end of lines (not including previous overlap)
	if overlapSize <= 0 {
		return nil
	}
	var result []lineObj
	overlapLen := 0
	for i := len(lines) - 1; i >= 0; i-- {
		lineLen := len(lines[i].text) + 1 // +1 for newline
		if overlapLen+lineLen > overlapSize && len(result) > 0 {
			break
		}
		result = append([]lineObj{lines[i]}, result...)
		overlapLen += lineLen
	}
	return result
}

func joinLines(lines []lineObj) string {
	parts := make([]string, len(lines))
	for i, l := range lines {
		parts[i] = l.text
	}
	return strings.Join(parts, "\n")
}

func sectionTextLen(section []lineObj) int {
	total := 0
	for i, l := range section {
		total += len(l.text)
		if i > 0 {
			total++ // newline between lines
		}
	}
	return total
}

// splitIntoSections splits file lines into logical sections based on extension.
func splitIntoSections(fileLines []string, ext string) [][]lineObj {
	switch {
	case ext == ".md":
		return splitMarkdown(fileLines)
	case ext == ".rst":
		return splitParagraphs(fileLines)
	case codeExtensions[ext]:
		return splitCode(fileLines)
	default:
		return splitParagraphs(fileLines)
	}
}

func splitMarkdown(fileLines []string) [][]lineObj {
	var sections [][]lineObj
	var current []lineObj
	inCodeBlock := false

	for i, line := range fileLines {
		lo := lineObj{lineNumber: i + 1, text: line}

		// Track fenced code blocks
		if strings.HasPrefix(strings.TrimLeft(line, " \t"), "```") {
			inCodeBlock = !inCodeBlock
			current = append(current, lo)
			continue
		}

		// Split on H2+ headers outside code blocks
		if !inCodeBlock && headerRegex.MatchString(line) && len(current) > 0 {
			sections = append(sections, current)
			current = []lineObj{lo}
			continue
		}

		// Split on blank lines (paragraph boundary) outside code blocks
		if !inCodeBlock && strings.TrimSpace(line) == "" && len(current) > 0 && hasNonBlank(current) {
			current = append(current, lo)
			sections = append(sections, current)
			current = nil
			continue
		}

		current = append(current, lo)
	}

	if len(current) > 0 {
		sections = append(sections, current)
	}
	return sections
}

func splitCode(fileLines []string) [][]lineObj {
	var sections [][]lineObj
	var current []lineObj

	for i, line := range fileLines {
		lo := lineObj{lineNumber: i + 1, text: line}

		if strings.TrimSpace(line) == "" && len(current) > 0 {
			hasNextNonBlank := false
			for j := i + 1; j < len(fileLines); j++ {
				if strings.TrimSpace(fileLines[j]) != "" {
					hasNextNonBlank = true
					break
				}
			}
			if hasNextNonBlank {
				sections = append(sections, current)
				current = []lineObj{lo}
				continue
			}
		}
		current = append(current, lo)
	}

	if len(current) > 0 {
		sections = append(sections, current)
	}
	return sections
}

func splitParagraphs(fileLines []string) [][]lineObj {
	var sections [][]lineObj
	var current []lineObj

	for i, line := range fileLines {
		lo := lineObj{lineNumber: i + 1, text: line}

		if strings.TrimSpace(line) == "" && len(current) > 0 {
			sections = append(sections, current)
			current = []lineObj{lo}
			continue
		}
		current = append(current, lo)
	}

	if len(current) > 0 {
		sections = append(sections, current)
	}
	return sections
}

func hasNonBlank(lines []lineObj) bool {
	for _, l := range lines {
		if strings.TrimSpace(l.text) != "" {
			return true
		}
	}
	return false
}
